### PR TITLE
[App] Provide better error message when an app is already running locally

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR improves the error message when a Lightning App is already running locally.

BEFORE:

```python
Your Lightning App is starting. This won't take long.
INFO: Your app has started. View it in your browser: http://127.0.0.1:7501/view
ERROR:    [Errno 48] error while attempting to bind on address ('127.0.0.1', 7501): address already in use
Task exception was never retrieved
future: <Task finished name='Task-1' coro=<Server.serve() done, defined at /Users/thomas/Documents/GitHub/lightning/.venv/lib/python3.9/site-packages/uvicorn/server.py:62> exception=SystemExit(1)>
Traceback (most recent call last):
  File "/Users/thomas/Documents/GitHub/lightning/.venv/lib/python3.9/site-packages/uvicorn/server.py", line 149, in startup
    server = await loop.create_server(
  File "uvloop/loop.pyx", line 1790, in create_server
OSError: [Errno 48] error while attempting to bind on address ('127.0.0.1', 7501): address already in use

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "uvloop/loop.pyx", line 474, in uvloop.loop.Loop._on_idle
  File "uvloop/cbhandles.pyx", line 83, in uvloop.loop.Handle._run
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/thomas/Documents/GitHub/lightning/.venv/lib/python3.9/site-packages/uvicorn/server.py", line 77, in serve
    await self.startup(sockets=sockets)
  File "/Users/thomas/Documents/GitHub/lightning/.venv/lib/python3.9/site-packages/uvicorn/server.py", line 159, in startup
    sys.exit(1)
SystemExit: 1
```

NOW:

```python
Your Lightning App is starting. This won't take long.
ERROR: A Lightning App is already running locally. HINT: Stop the previous App or run in the cloud with `--cloud`.
INFO: Your Lightning App is being stopped. This won't take long.
INFO: Your Lightning App has been stopped successfully!
``` 

Fixes #\<issue_number>

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @akihironitta

### CODE REVIEW

**Review by ethanwharris — APPROVED** [review]

File: src/lightning_app/runners/multiprocess.py [hunk 1]
```diff
@@ -63,6 +66,13 @@ def dispatch(self, *args: Any, on_before_run: Optional[Callable] = None, **kwarg
             storage_orchestrator.start()
 
             if self.start_server:
+                if check_port_already_used(self.port):
+                    logger.error(
+                        "A Lightning App is already running locally. "
+                        "HINT: Stop the previous App or run in the cloud with `--cloud`."
```
Comment (ethanwharris) [inline src/lightning_app/runners/multiprocess.py:72]: Would maybe change this to:
```suggestion
                        "Stop the existing App or run in the cloud with `--cloud`."
```

File: src/lightning_app/utilities/network.py [hunk 2]
```diff
@@ -27,6 +28,22 @@ def find_free_network_port() -> int:
...
+
+    try:
+        s.bind(("127.0.0.1", port))
+    except OSError as e:
+        if e.errno == errno.EADDRINUSE:
+            used = True
+
+    s.close()
+
+    return used
```
Comment (Borda) [inline src/lightning_app/utilities/network.py:44]:
```suggestion
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sct:
        try:
            sct.bind(("127.0.0.1", port))
        except OSError as ex:
            if ex.errno == errno.EADDRINUSE:
                return True

    return False
```

**Review by Borda — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

File: src/lightning_app/utilities/network.py [hunk 3]
```diff
@@ -27,6 +28,17 @@ def find_free_network_port() -> int:
     return port
 
 
+def check_port_already_used(port: int) -> int:
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sct:
+        try:
+            sct.bind(("127.0.0.1", port))
+        except OSError as ex:
+            if ex.errno == errno.EADDRINUSE:
+                return True
```
Comment (awaelchli) [inline src/lightning_app/utilities/network.py:37]:
```suggestion
            return ex.errno == errno.EADDRINUSE
```

**Comment by awaelchli** [comment]
@tchaton There is an alternative approach here: #15819 Mind taking a look?

**Comment by Borda** [comment]
@tchaton @awaelchli so which direction we shall take?

**Comment by awaelchli** [comment]
@tchaton this might be solved now by #15819

**Pull request closed without merging**